### PR TITLE
Feat/min bitmap

### PR DESCRIPTION
**Pull request by detectivekim**
Implement `MinBitMap`.

### CODE REVIEW

File: contracts/MinBitMap.sol [hunk 1]
```diff
@@ -0,0 +1,82 @@
+// SPDX-License-Identifier: UNLICENSED
```
Comment (JhChoy) [inline contracts/MinBitMap.sol]: Check this.

Comment (detectivekim) [reply contracts/MinBitMap.sol]: Resolved.

File: test/foundry/MinBitMap.t.sol [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+        assertFalse(testWrapper.isEmpty(), "HAS_TO_BE_OCCUPIED");
+        while (!testWrapper.isEmpty()) {
+            _min = testWrapper.root();
+            assertTrue(testWrapper.has(_min), "BITMAP_HAS_ROOT");
+            testWrapper.pop();
+            assertFalse(testWrapper.has(_min), "ROOT_HAS_BEEN_POPPED");
+            if (testWrapper.isEmpty()) break;
+            assertGt(testWrapper.root(), _min, "ROOT_HAS_TO_BE_MIN");
+        }
+    }
```
Comment (JhChoy) [inline test/foundry/MinBitMap.t.sol]: We need the tests below.
- A test about calling pop() when the heap is empty.
- A test when pushing the same number.
- A test when calling root() when the heap is empty.

Comment (detectivekim) [reply test/foundry/MinBitMap.t.sol]: Resolved.

File: contracts/Heap.sol [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+pragma solidity ^0.8.0;
+
+import "./SignificantBit.sol";
+
+library Heap {
+    using SignificantBit for uint256;
+
+    error HeapError(uint256 errorCode);
+    uint256 public constant EMPTY_ERROR = 0;
+    uint256 public constant ALREADY_EXISTS_ERROR = 1;
```
Comment (JhChoy) [inline contracts/Heap.sol]: Let's divide errors into two.

Comment (detectivekim) [reply contracts/Heap.sol]: Resolved.

File: contracts/Heap.sol [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+            revert AlreadyExistsError();
+        }
+
+        heap[b0b1] = b2Bitmap | mask;
+        if (b2Bitmap == 0) {
+            mask = 1 << (b0b1 & 0xff);
+            uint256 b0 = b0b1 >> 8;
+            heap[B0_BITMAP_KEY] = heap[B0_BITMAP_KEY] | (1 << b0);
+            b0 = ~b0; // TODO add comment
+            heap[b0] = heap[b0] | mask;
```
Comment (JhChoy) [inline contracts/Heap.sol]: We can reduce unnecessary sload when b1Bitmap is not empty like current `pop` logic.
```suggestion
            mask = 1 << (b0b1 & 0xff);
            uint256 b1BitmapKey = ~(b0b1 >> 8);
            uint256 b1Bitmap = heap[b1BitmapKey];

            heap[b1BitmapKey] = b1Bitmap | mask;
            if (b1Bitmap == 0) {
                 heap[B0_BITMAP_KEY] = heap[B0_BITMAP_KEY] | (1 << ~b1BitmapKey);
            }
```

File: contracts/Heap.sol [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+        (uint256 b0b1, uint256 b2) = _split(value);
+        uint256 b2Bitmap = (MAX_UINT_256_MINUS_1 << b2) & heap[b0b1];
+        if (b2Bitmap == 0) {
+            uint256 b0 = b0b1 >> 8;
+            uint256 b1Bitmap = (MAX_UINT_256_MINUS_1 << (b0b1 & 0xff)) & heap[~b0];
+            if (b1Bitmap == 0) {
+                uint256 b0Bitmap = (MAX_UINT_256_MINUS_1 << b0) & heap[B0_BITMAP_KEY];
+                if (b0Bitmap == 0) {
+                    revert EmptyError();
+                }
```
Comment (JhChoy) [inline contracts/Heap.sol]: In my opinion, returning something exceptional is better than reverting when there is no greater value in the heap.
How about returning `value` itself?

**Review by JhChoy — APPROVED** [review]

**Pull request merged**